migrations/dynamodb-migrations/2021-08-04T10:06:50Z-migrate-sampleids-to-experiment.py

```diff
@@ -0,0 +1,156 @@
+
+environment = 'production'
+
+experiments_table = f'experiments-{environment}'
+projects_table = f'projects-{environment}'
+samples_table = f'samples-{environment}'
+
+experiments = client.scan(
+    TableName=experiments_table,
+    ProjectionExpression="experimentId, createdAt, projectId",
+)['Items']
+
+## >> Uncomment lines below to test with a single experiment
+# test_experiment_id = "b947598cc38860660a82ffd508e2b068"
+
+# experiments = client.get_item(
+#     TableName=experiments_table,
+#     Key={"experimentId": { "S" : test_experiment_id }},
+#     ProjectionExpression="experimentId, createdAt, projectId"
+# )['Item']
+# experiments = [experiments]
+
+## << Uncomment lines above to test
+
+# function to create gem2s hash
+def create_gem2s_hash(experiment, project, samples):
+
+    organism = None # default value
+
+    if experiment['meta']['M']['organism'].get('S'):
+        organism = experiment['meta']['M']['organism']['S']
+
+    # Filter 'ids' key which is present in older samples object
+    sample_ids = [sample_id for sample_id in samples['M'] if sample_id != 'ids']
+
+    sample_names = []
+
+    # Sample names
+    for sample_id in sample_ids:
+        sample_names.append(samples['M'][sample_id]['M']['name']['S'])
+
+    # Sample IDS and names are sorted to be invariant
+    sample_ids.sort()
+    sample_names.sort()
+
+    input_type = experiment["meta"]['M']["type"].get('S', '10x')
+
+    hash_params = {
+        "organism": organism,
+        "input": {"type" : input_type},
+        "sampleIds": sample_ids,
+        "sampleNames": sample_names,
+    }
+
+    metadata_values = OrderedDict()
+
+    if len(project['M']['metadataKeys']['L']) > 0:
+
+        metadata_keys = [metadata['S'] for metadata in project['M']['metadataKeys']['L']]
+
+        for key in metadata_keys:
+            # Replace '-' in key to '_'if
+            sanitizedKey = key.replace('-', '_')
+
+            for sample_id in sample_ids:
+
+                metadata_value = "N.A." # default metadata value
+
+                if samples['M'][sample_id]['M']['metadata']['M'].get(key):
+                    metadata_value = samples['M'][sample_id]['M']['metadata']['M'][key]['S']
+
+                if not metadata_values.get(sanitizedKey):
+                    metadata_values[sanitizedKey] = []    
+
+                metadata_values[sanitizedKey].append(metadata_value)
+
+        hash_params['metadata'] = metadata_values
+
+    hash_params_string = json.dumps(hash_params).replace(", ", ",").replace(": ", ":").encode('utf-8')
+
+    return hashlib.sha1(hash_params_string).hexdigest()
+
+
+print(f"=== Doing migration for PR 408 ===")
+
+for experiment in experiments:
+
+    try:
+        experiment_id = experiment['experimentId']
+        print(f"\n=========")
+        print(f"Preparing migration for experiment {experiment_id['S']}")
+
+        if not experiment.get('createdAt') : print('Project does not contain createdAt, inserting current timestamp')
```

Comment:
the default date value if it is missing should be the `createdDate` of the project. no need to change it because we have few users and so far we aren't relying on this metrics.